### PR TITLE
Add extension 'DrawNodeEx' aka 'DrawNode v2' to Axmol Engine => Initial PR

### LINKED ISSUE #1567: Extension DrawNodeEx ‐ a new kind of sprite class (*experimental phase*)
FIRST:
Because of possible side effects with `class DrawNode` the extension `DrawNodeEx` aka `DrawNode v2` is still in the experimental phase and may encounter bugs or stability or compatibility issues when used.

DrawNodeEx is not only a class for drawing primitives, it's more: 
- It's `DrawNode v2`
- It's a kind of sprite created from drawing primitives!

Features:
- `Individual line width`
```
    setDNLineWidth();
    getDNLineWidth();

    Default value: `1.0f`
```

- `Scale/Rotation/Position`
```
Scale:
    setDNScale()
    setDNScaleX()
    setDNScaleY()
Rotation:
    setDNRotation()
    getDNRotation()
    setDNCenter()
    getDNCenter()
Position:
    setDNPosition()
    getDNPosition()
```
- `general`
```
ensureDNValues()
restoreDNValues()
resetDNValues()
```
- Planed APIs:
```
enableDNTransform()
getDNSprite()
explodeDNSprite()
transformDNSprite()
morphDNSprite()
createAnimDNSprite()
```

- [x] Add initial `DrawNode v2` as extension `DrawNodeEx` to axmol as experimental 
- [x] `DrawNode` API compatible
- [x] Add individual line width for drawing primitives 
- [x] Scale drawing primitives without changing the individual line width.
- [x] Rotate drawing primitives around a customized center 
- Useful tests
     - [ ] funktional
     - [ ] performance/robustness
     - [ ] compare `DrawNode` vs `DrawNode v2`
- [ ] Code Cleanup
     - [ ] Remove/Add some comments
     - [ ] Reduce compiler warnings
- [ ] Documentation, Wiki
     - [ ] Info about AX_TRANSPARENT

New drawMethods:
- drawStar()
- drawSolidStar()

image

### PR DESCRIPTION
**Pull request by aismann**
Which branch your pull-request should merge into?

- `dev`: Current 2.x BugFixs or Features

## Describe your changes
See "Class DrawNodeExt: All info here (*experimental phase*) #1567"

## Issue ticket number and link
#1567

### For each PR
-  [x] I have performed a self-review of my code.
       
   Optional:
   -  [ ] I have checked readme and add important infos to this PR.
   -  [ ] I have added/adapted some tests too.
          
### For core/new feature
-  [x] I have checked readme and add important infos to this PR.
-  [x] I have added thorough tests.

### CODE REVIEW

**Comment by aismann** [comment]
@halx99 
Please ignore 'Codacy Static Code Analysis'  C-style pointer casting is here like on class DrawNode

**Comment by aismann** [comment]
EDIT: @halx99 , can you have a look what is wrong (now its always shown)?


@halx99 
I want it optional like: 
```c++
#if __has_include("EffekseerForCocos2d-x.h")
#pragma message("The optional extension Effekseer is enabled.")
        addTest("Effekseer", []() { return new EffekseerTests(); });
#endif
```

But its dimmed: What is missed?
```c++
#if __has_include("DrawNodeExTest.h")
#    pragma message("The optional extension DrawNodeEx is enabled.")
        addTest("DrawNodeEx *experimental*", []() { return new DrawNodeExTests(); });
#endif
```

**Comment by aismann** [comment]
@halx99 
Please merge this initial version to Axmol Engine 2.1.0

**Comment by aismann** [comment]
Finish now ;)

@halx99 
New tester will comes later, please merge this initial version of DrawNodeEx

**Comment by aismann** [comment]
replace it with experimental

File: extensions/DrawNodeEx/DrawNodeEx.cpp [hunk 1]
```diff
@@ -0,0 +1,1213 @@
...
+    }
+
+    _customCommandTriangle.updateVertexBuffer(triangles, _bufferCountTriangle * sizeof(V2F_C4B_T2F),
+                                              vertex_count * sizeof(V2F_C4B_T2F));
+    _bufferCountTriangle += vertex_count;
+    _customCommandTriangle.setVertexDrawInfo(0, _bufferCountTriangle);
+    _dirtyTriangle = true;
+}
+
+inline void DrawNodeEx::_drawPoly(const Vec2* poli,
```
Comment (halx99) [inline extensions/DrawNodeEx/DrawNodeEx.cpp]:
```suggestion
void DrawNodeEx::_drawPoly(const Vec2* poli,
```

Comment (aismann) [reply extensions/DrawNodeEx/DrawNodeEx.cpp]: @halx99 
`inline` I thought makes a better performance, isnt it?
Avoiding the cost of a function call

Comment (halx99) [reply extensions/DrawNodeEx/DrawNodeEx.cpp]: No， The code large, not recommnd add inline keyword, and the compiler actually ignore the inline mark when a function to large or logic too complex, never mark function as inline when contains condition check, loop statments

**Comment by aismann** [comment]
@halx99 
Ready for v2.1.0-rc8

**Comment by aismann** [comment]
@halx99 
I think it's make sense to publish Axmol  v2.1.0-rc8 with extension DrawNodeEx for everybody testing.
Finding mistakes/wishes/bugs before Axmol v2.1.0 is released is wished

**Comment by halx99** [comment]
> @halx99 I think it's make sense to publish Axmol v2.1.0-rc8 with extension DrawNodeEx for everybody testing. Finding mistakes/wishes/bugs before Axmol v2.1.0 is released is wished

sure, will do

**Pull request merged**